src/main/java/org/opensearch/ad/feature/CompositeRetriever.java

```diff
@@ -0,0 +1,346 @@
+        long expirationEpochMs,
+        Clock clock,
+        Settings settings,
+        int maxEntitiesPerInterval,
+        int pageSize
+    ) {
+        this.dataStartEpoch = dataStartEpoch;
+        this.dataEndEpoch = dataEndEpoch;
+        this.anomalyDetector = anomalyDetector;
+        this.xContent = xContent;
+        this.client = client;
+        this.totalResults = 0;
+        this.maxEntities = maxEntitiesPerInterval;
+        this.pageSize = pageSize;
+        this.expirationEpochMs = expirationEpochMs;
+        this.clock = clock;
+    }
+
+    // a constructor that provide default value of clock
+    public CompositeRetriever(
+        long dataStartEpoch,
+        long dataEndEpoch,
+        AnomalyDetector anomalyDetector,
+        NamedXContentRegistry xContent,
+        Client client,
+        long expirationEpochMs,
+        Settings settings,
+        int maxEntitiesPerInterval,
+        int pageSize
+    ) {
+        this(
+            dataStartEpoch,
+            dataEndEpoch,
+            anomalyDetector,
+            xContent,
+            client,
+            expirationEpochMs,
+            Clock.systemUTC(),
+            settings,
+            maxEntitiesPerInterval,
+            pageSize
+        );
+    }
+
+    public void start(ActionListener<Page> listener) {
+        try {
+            RangeQueryBuilder rangeQuery = new RangeQueryBuilder(anomalyDetector.getTimeField())
+                .gte(dataStartEpoch)
+                .lt(dataEndEpoch)
+                .format("epoch_millis");
+
+            BoolQueryBuilder internalFilterQuery = new BoolQueryBuilder().filter(anomalyDetector.getFilterQuery()).filter(rangeQuery);
+
+            // multiple categorical fields are supported
+            CompositeAggregationBuilder composite = AggregationBuilders
+                .composite(
+                    AGG_NAME_COMP,
+                    anomalyDetector
+                        .getCategoryField()
+                        .stream()
+                        .map(f -> new TermsValuesSourceBuilder(f).field(f))
+                        .collect(Collectors.toList())
+                )
+                .size(pageSize);
+            for (Feature feature : anomalyDetector.getFeatureAttributes()) {
+                AggregatorFactories.Builder internalAgg = ParseUtils
+                    .parseAggregators(feature.getAggregation().toString(), xContent, feature.getId());
+                composite.subAggregation(internalAgg.getAggregatorFactories().iterator().next());
+            }
+
+            // In order to optimize the early termination it is advised to set track_total_hits in the request to false.
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+                .query(internalFilterQuery)
+                .size(0)
+                .aggregation(composite)
+                .trackTotalHits(false);
+
+            Page page = new Page(null, searchSourceBuilder, null);
```

Comment:
Does the iterable fit to this use case? e.g. class IterablePage extends Iterable<Result>

Reply:
You meant https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html ?  I need parameters in the next method.  Does not seem fit.

Reply:
I tried to separate page and pageIterator.  Please check the new commit.  Also, cannot extend JDK interface since the method signature next() does not match my need.  I need to return result using async listener callback.  Iterator.next() in JDK is a synchronous method.